Feature/add configurable fee

OliverNChalk opened the pull request:
The goal of this PR is to replace the hard-coded 5bp / 1/6th fee with a variable LP fee that can be set from 0 - 5000 bp (50%). I also add a function that controls the swap fee within a range from 0-200 bp.

OliverNChalk committed:
UniswapV2Pair: Initial draft of variable fee

OliverNChalk committed:
Src: More scribbles and experiments with a variable fee, fee formula needs rewriting

OliverNChalk committed:
Src: Add example CalcFee.sol contract, needs care around overflows / accuracy

OliverNChalk committed:
CalcFee: Add method that does not overflow due to kLast

OliverNChalk committed:
UniswapV2Pair: Integrate new fee calc, pending testing and refactor

OliverNChalk committed:
UniswapV2Factory: Add functions to manage swapFee, platformFee, and recoverer

OliverNChalk committed:
UniswapV2Pair: Allow recoverer to be revoked, disabling recovery

OliverNChalk committed:
Core: Cleanup for review, have not run tests or tried building outside of RemixIDE

svechinsky commented on uniswap-v2-core/contracts/CalcFee.sol:
```diff
@@ -0,0 +1,68 @@
+        uint256 _scaledTargetOwnership = _scaledMultiplier * protocol_fee / FEE_ACCURACY;   // Assert: < uint144 during maths, ends < uint128
+
+        // NOTE: circulating_shares in uniV2 is initialized by taking the sqrt(token_a * token_b) ensuring that as a number it will be larger
+        // than the large of the two numbers. So if token_a has 18 decimals and token_b has 10 decimals we can expect circulating_shares / liquidity
```
Than the smaller of the two numbers no?

OliverNChalk replied:
Correct, I will fix the wording. This trick basically helps us reason about overflow scenarios

svechinsky commented on uniswap-v2-core/contracts/UniswapV2Factory.sol:
```diff
@@ -2,18 +2,28 @@ pragma solidity =0.5.16;
 import './UniswapV2Pair.sol';
+import './libraries/Ownable.sol';
+
+contract UniswapV2Factory is IUniswapV2Factory, Ownable {
```
Weird that uniswap didn't use this?
Did they need to save so much on the size of the contract?

OliverNChalk replied:
I think it was for optics, `Ownable` goes against crypto ethos, even though it doesn't grant custodial powers

svechinsky commented on uniswap-v2-core/contracts/UniswapV2Factory.sol:
```diff
@@ -37,13 +47,36 @@ contract UniswapV2Factory is IUniswapV2Factory {
+        IUniswapV2Pair(_pair).setPlatformFee(_platformFee);
+    }
+    
+    function setRecovererForPair(address _pair, address _recoverer) external onlyOwner {
```
No timelock on this?

OliverNChalk replied:
This function _SHOULD_ only be able to recover non-core assets (i.e. it should only allow us to withdraw assets that would have otherwise been stuck). If we want to make this timelocked, we can make the owner a timelock contract and thus timelock the management of the recoverer.

We could even make the owner a contract that hardcodes the recoverer for all time (if we're that confident we don't want to change it).

svechinsky replied:
You're right, missed it

svechinsky commented on uniswap-v2-core/contracts/CalcFee.sol:
```diff
@@ -0,0 +1,68 @@
+// SPDX-License-Identifier: UNLICENSE// SPDX-License-Identifier: UNLICENSED
+pragma solidity >=0.7.0 <0.9.0;
+
+contract CalcFee {
```
Where do you use this code?

OliverNChalk replied:
This was an example file to let people deploy and test the minting function in isolation. This code made it into UniswapV2Pair.sol under `_calcFee`

svechinsky commented on uniswap-v2-core/contracts/UniswapV2Pair.sol:
```diff
@@ -85,20 +118,31 @@ contract UniswapV2Pair is IUniswapV2Pair, UniswapV2ERC20 {
+            uint _sqrtOldK = Math.sqrt(kLast); // gas savings
+        
+            if (_sqrtOldK != 0){
+                if (_sqrtNewK > _sqrtOldK) {
```
No fee on exit?

OliverNChalk replied:
`_mintFee` is called inside `mint` and `burn`, UniswapV2's enter and exit functions. `sqrtNew` sqrtOld` should be taken _BEFORE_ processing the mint and burn and `sqrtOld` should be saved _AFTER_ processing the mint and burn. This means we only measure the fee performance between liquidity additions and removals and we don't charge any fees for inserting or removing liquidity.

svechinsky reviewed (commented):
Added comments/questions

leodsoft committed:
Align solidity version

leodsoft commented on uniswap-v2-core/contracts/UniswapV2Pair.sol:
```diff
@@ -177,9 +221,10 @@ contract UniswapV2Pair is IUniswapV2Pair, UniswapV2ERC20 {
-        uint balance0Adjusted = balance0.mul(1000).sub(amount0In.mul(3));
-        uint balance1Adjusted = balance1.mul(1000).sub(amount1In.mul(3));
-        require(balance0Adjusted.mul(balance1Adjusted) >= uint(_reserve0).mul(_reserve1).mul(1000**2), 'UniswapV2: K');
+        // TODO: Check if 1,000 -> 10,000 introduces overflow possibility
```
This will just reduce max balances by 14bit instead of 10bit, to safely perform the check; I note you had a theoretical max-balance of 128bit discussed in calcfee.sol comments ==> this would reduce that limit to 114bit.

leodsoft replied:
114bit still provides an enormous balance - but what is the expected range in the market? (Should the contract need to manage max balances?)

leodsoft replied:
Reading the white paper - this should be covered already. Uni swap v2 only supports 112bit balances anyway, so all clear. Higher balances will cause swap() to fail, and a user can use skim() to recover excess from the LP.

OliverNChalk replied:
Great, resolved

leodsoft commented on uniswap-v2-core/contracts/CalcFee.sol:
```diff
@@ -0,0 +1,68 @@
+// SPDX-License-Identifier: UNLICENSE// SPDX-License-Identifier: UNLICENSED
+pragma solidity >=0.7.0 <0.9.0;
```
should align with solidity version of other contracts (or we can drop CalcFee)

leodsoft commented on uniswap-v2-core/contracts/UniswapV2Pair.sol:
```diff
@@ -85,20 +118,31 @@ contract UniswapV2Pair is IUniswapV2Pair, UniswapV2ERC20 {
+    function _calcFee(uint _sqrtNewK, uint _sqrtOldK, uint _platformFee, uint _circulatingShares) internal pure returns (uint _sharesToIssue) {
+        // Assert newK & oldK        < uint112
+        // Assert _circulatingShares < FEE_ACCURACY
+        // Assert _circulatingShares < uint112
```
Should we consider enforcing this asserts - or just leave as guides for the reader?
Also - why _circulatingShares < FEE_ACCURACY?

leodsoft replied:
(confirmed with @OliverNChalk that this should be: "Assert _platformFee < FEE_ACCURACY")

leodsoft commented on uniswap-v2-core/contracts/libraries/Ownable.sol:
```diff
@@ -0,0 +1,53 @@
+    function renounceOwnership() public onlyOwner {
+        emit OwnershipTransferred(_owner, address(0));
+        _owner = address(0);
+    }
```
Might be good to remove this ability: although a natural part of Ownable, dropping ownership will probably not make sense in this context.

leodsoft committed:
Corrected protocol_fee assertion / comment

leodsoft committed:
Remove renounceOwnership

OliverNChalk committed:
V2Core: Fix comment issue, rename platform_fee to protocol_fee

OliverNChalk commented on uniswap-v2-core/contracts/UniswapV2Pair.sol:
```diff
@@ -2,19 +2,31 @@ pragma solidity =0.5.16;
 
     uint public constant MINIMUM_LIQUIDITY = 10**3;
     bytes4 private constant SELECTOR = bytes4(keccak256(bytes('transfer(address,uint256)')));
+    uint256 public constant SQUARED_ACCURACY = 100_000_000_000_000_000_000_000_000_000_000_000_000; // 100000000000000000000000000000000000000
```
This value is currently the power of 10 right before breaching 2^128. We could look at slightly breaching 2^128, such that we guarantee no precision loss on division of two numbers if both are < 2^128. This depends on what the size of the balances we're dividing are. If they're less than the current `SQUARED_ACCURACY` - no change will be required.

OliverNChalk replied:
```
    function _mintFee(uint112 _reserve0, uint112 _reserve1) private returns (bool feeOn) {
            uint _sqrtNewK = Math.sqrt(uint(_reserve0).mul(_reserve1));
            uint _sqrtOldK = Math.sqrt(kLast); // gas savings
    }
```

So the inputs will never be above uint112 (assumng sqrt(uint112 * uint112) < uint112). `SQUARED_ACCURACY` should likely be increased to just below `uint256(-1)` so that we can get 256 - 112 bits of accuracy in the division result? `ACCURACY` would then be just below uint128 but still safe because it is only ever multiplied with a uin112

leodsoft replied:
With the accuracy for the calc now increased and proven with base-line test cases, we can likely close out this review point: though further platformFee testing will help confirm it further.

OliverNChalk replied:
Agreed, `SQUARED_ACCURACY` is now at the maximum value before overflowing and no obvious improvements come to mind.

OliverNChalk commented on uniswap-v2-core/contracts/UniswapV2Pair.sol:
```diff
@@ -35,6 +47,27 @@ contract UniswapV2Pair is IUniswapV2Pair, UniswapV2ERC20 {
+    }
+
+    function setSwapFee(uint _swapFee) external onlyFactory {
+        require(_swapFee > MIN_SWAP_FEE && _swapFee < MAX_SWAP_FEE, "Vexchange: INVALID_SWAP_FEE");
```
Align this `require` check with the factory. Either make the factory all in one line, or split this `require` over two lines

OliverNChalk commented on uniswap-v2-core/contracts/UniswapV2Pair.sol:
```diff
@@ -35,6 +47,27 @@ contract UniswapV2Pair is IUniswapV2Pair, UniswapV2ERC20 {
+        _;
+    }
+
+    function setRecoverer(address _recoverer) external onlyFactory {
```
Should order functions the same as factory:
 1. Set swap
 2. Set platform
 3. Set recoverer

OliverNChalk replied:
Resolved

OliverNChalk commented on uniswap-v2-core/contracts/UniswapV2Pair.sol:
```diff
@@ -85,20 +118,31 @@ contract UniswapV2Pair is IUniswapV2Pair, UniswapV2ERC20 {
+        // Assert _platformFee       < FEE_ACCURACY
+        // Assert _circulatingShares < uint112
+
+        uint256 _scaledGrowth = (_sqrtNewK * ACCURACY) / _sqrtOldK;                      // ASSERT: < UINT256
+        uint256 _scaledMultiplier = ACCURACY - (SQUARED_ACCURACY / _scaledGrowth);          // ASSERT: < UINT128
+        uint256 _scaledTargetOwnership = _scaledMultiplier * _platformFee / FEE_ACCURACY; // ASSERT: < UINT144 during maths, ends < UINT128
```
Use SafeMaths here, confirm we can rescue funds by turning platformFee off and thus not entering this function

OliverNChalk replied:
Should be implemented, we can check SafeMath usage codebase wide once tests are more developed

OliverNChalk commented on uniswap-v2-core/contracts/UniswapV2Pair.sol:
```diff
@@ -85,20 +118,31 @@ contract UniswapV2Pair is IUniswapV2Pair, UniswapV2ERC20 {
-                    if (liquidity > 0) _mint(feeTo, liquidity);
+            address feeTo = IUniswapV2Factory(factory).feeTo();
+            uint _sqrtNewK = Math.sqrt(uint(_reserve0).mul(_reserve1));
+            uint _sqrtOldK = Math.sqrt(kLast); // gas savings
```
Confirm no stack too deep issue from these new variables

leodsoft replied:
This should be verified through solc compilation I think? Or do we need to cover with additional testing as well?

OliverNChalk replied:
Ahh, yes - you appear to be right. This should absolutely be caught at compile time and online discourse seems to indicate that. If not then, im sure we hit all paths in tests?

OliverNChalk commented on uniswap-v2-core/contracts/UniswapV2Pair.sol:
```diff
@@ -85,20 +118,31 @@ contract UniswapV2Pair is IUniswapV2Pair, UniswapV2ERC20 {
-                    uint denominator = rootK.mul(5).add(rootKLast);
-                    uint liquidity = numerator / denominator;
-                    if (liquidity > 0) _mint(feeTo, liquidity);
+            address feeTo = IUniswapV2Factory(factory).feeTo();
```
Decide if we leave as `feeTo` because that's what it was, or move to `_feeTo` because its a local variable. Unsure of what Uniswap's convention is and whether they themselves are following it

leodsoft replied:
Unfortunately - just from reading our initial check-in (I didn't check original source) - if they are consistent, I have a feeling that uniswap might only use _ when there is an internal member var of the same name... either that or they are just inconsistent.

leodsoft replied:
We should discuss - despite their std, I think it would be preferable to prefix all args & local vars...

OliverNChalk replied:
Yep it does seem like `_` is more of a hack to let them redeclare variables locally. So either because they want a copy of the old value or they want to avoid loading the old value from storage too many times.

OliverNChalk commented on uniswap-v2-core/contracts/UniswapV2Pair.sol:
```diff
@@ -35,6 +47,27 @@ contract UniswapV2Pair is IUniswapV2Pair, UniswapV2ERC20 {
+        recoverer = _recoverer;
+    }
+
+    function setPlatformFee(uint _platformFee) external onlyFactory {
```
Should these be `_variable` given external method? Maybe drop the `_`

leodsoft replied:
I think this is consistent with the rest of uniswap (apply _ only to provide uniqueness when required) - so I'd suggest we leave as-is for the moment.

OliverNChalk reviewed (commented):
Reviewed the code and added some comments

OliverNChalk committed:
V2Factory: Add the new methods and properties to interface

leodsoft committed:
Aligned factory interface, correct test-cases for Factory (all should be passing at this point).

OliverNChalk committed:
Src: Sync factory interfaces in core and periphery

OliverNChalk committed:
Tools: Add install-deps step to ensure consistent environment for tests

OliverNChalk committed:
Src: Align IUniswapV2Pool.sol across core and periphery

OliverNChalk committed:
Config: Change compiler versions to istanbul, set chaintags to testnet

OliverNChalk committed:
Misc: Update README.md and change one remaining usage of istanbul option

OliverNChalk committed:
Merge: More alignment with Kenneth's version

OliverNChalk committed:
Core: Add events on swapFee, platformFee, recoverer, and feeTo changes

OliverNChalk committed:
V2Pair: Use SafeMath for _calcFee

leodsoft committed:
Minor improvements

Parmetrise package mgr for deps - may be temporary, but yarn is failing on my macos env;
Separate package mgmt from primary make usecase (make test);
Added ‘make all’ to install, compile and test - as default (‘make’).

leodsoft committed:
Correctly initialise pair with new fee parameters

leodsoft committed:
Align basic tests to original fee structure

leodsoft committed:
Align IUniswapV2Pair interface in Periphery

leodsoft committed:
Correctly initialise pairs in periphery tests (simple-default fee settings)

leodsoft committed:
Correct min/max assertions

leodsoft committed:
Working base test cases with new variable fees

These tests so far only cover fees set to original uniswap v2 values.

OliverNChalk committed:
Environment: Standardize around using npm over yarn

leodsoft committed:
Align periphery interfaces with core

leodsoft committed:
Rename feeTo to platformFeeTo & associated functions

leodsoft committed:
Merge branch 'feature/add-configurable-fee' of https://github.com/OliverNChalk/vexchange-contracts into feature/add-configurable-fee

leodsoft committed:
Update periphery for feeTo name change (to platformFee)

leodsoft committed:
Add initial swapFee and platform fee to createPair event

leodsoft committed:
Fix VM exceptions (update hard-coded UniswapV2Pair init code hash)

leodsoft committed:
Clean periphery tests.

leodsoft committed:
Fix install makefile step

leodsoft committed:
Align with Pair

leodsoft committed:
Replace use of js number with ethers/utils BigNumber

leodsoft committed:
Add utility method to provide Pair init-code.

OliverNChalk committed:
Config: Standardize scripts to use npm or globals

leodsoft committed:
Fix whitespace

leodsoft committed:
Adding unit test for pair init-code validation

OliverNChalk committed:
UniswapV2Pair: Increase SQUARED_ACCURACY to max uint256, ACCURACY updated in line

OliverNChalk committed:
Make: Replace npm install with npm ci

leodsoft committed:
Aligned tests for increased accuracy

leodsoft committed:
Update pair init-code.

OliverNChalk reviewed (commented):
Good to go, there are some whitespace issues but going to leave these as they can always come in during the branding refactor if they really stress us (me)
